pipenv/core.py

```diff
@@ -2184,44 +2184,66 @@ def inline_activate_virtualenv():
+    args = list(args or [])
+    # Script was found…
+    if command in project.scripts:
+        prefix_argv = project.scripts[command]
+    else:
+        # Separate out things that were passed in as a string.
+        prefix_argv = list(command.split())
+    executable, prefix_argv = prefix_argv[0], prefix_argv[1:]
+    for __c in reversed(prefix_argv):
+        args.insert(0, __c)
+    return executable, args
+
+
 def do_run(command, args, three=None, python=False):
     # Ensure that virtualenv is available.
     ensure_project(three=three, python=python, validate=False)
     load_dot_env()
-    # Script was found…
-    if command in project.scripts:
-        command = ' '.join(project.scripts[command])
-    # Separate out things that were passed in as a string.
-    _c = list(command.split())
-    command = _c.pop(0)
-    if _c:
-        args = list(args)
-        for __c in reversed(_c):
-            args.insert(0, __c)
+    executable, args = _construct_run_command(project, command, args)
     # Activate virtualenv under the current interpreter's environment
     inline_activate_virtualenv()
     # Windows!
     if os.name == 'nt':
         import subprocess
 
         p = subprocess.Popen(
-            [command] + list(args), shell=True, universal_newlines=True
+            [executable] + list(args), shell=True, universal_newlines=True
         )
         p.communicate()
         sys.exit(p.returncode)
     else:
-        command_path = system_which(command)
+        command_path = system_which(executable)
         if not command_path:
-            click.echo(
-                '{0}: the command {1} could not be found within {2} or Pipfile\'s {3}.'
-                ''.format(
-                    crayons.red('Error', bold=True),
-                    crayons.red(command),
-                    crayons.normal('PATH', bold=True),
-                    crayons.normal('[scripts]', bold=True),
-                ),
-                err=True,
-            )
+            if command in project.scripts:
+                click.echo(
+                    '{0}: the command {1} (from {2}) could not be found within {3}.'
+                    ''.format(
+                        crayons.red('Error', bold=True),
+                        crayons.red(executable),
+                        crayons.normal(command, bold=True),
+                        crayons.normal('PATH', bold=True),
+                    ),
+                    err=True,
+                )
+            else:
+                click.echo(
+                    '{0}: the command {1} could not be found within {2} or Pipfile\'s {3}.'
+                    ''.format(
+                        crayons.red('Error', bold=True),
+                        crayons.red(command),
+                        crayons.normal('PATH', bold=True),
+                        crayons.normal('[scripts]', bold=True),
+                    ),
+                    err=True,
+                )
```

Comment:
These are good, but why not Windows?

Reply:
Why? Windows also uses `PATH` for lookup, and the error message applies as well on Windows to me.

Reply:
Ah, I see what you mean, sorry. The Windows branch needs the same error messages as well. Right.